core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java

```diff
@@ -0,0 +1,603 @@
+    }
+
+    public InnerHitBuilder setVersion(boolean version) {
+        this.version = version;
+        return this;
+    }
+
+    public boolean isTrackScores() {
+        return trackScores;
+    }
+
+    public InnerHitBuilder setTrackScores(boolean trackScores) {
+        this.trackScores = trackScores;
+        return this;
+    }
+
+    public List<String> getFieldNames() {
+        return fieldNames;
+    }
+
+    public InnerHitBuilder setFieldNames(List<String> fieldNames) {
+        this.fieldNames = fieldNames;
+        return this;
+    }
+
+    public List<String> getFieldDataFields() {
+        return fieldDataFields;
+    }
+
+    public InnerHitBuilder setFieldDataFields(List<String> fieldDataFields) {
+        this.fieldDataFields = fieldDataFields;
+        return this;
+    }
+
+    public InnerHitBuilder addFieldDataField(String field) {
+        if (fieldDataFields == null) {
+            fieldDataFields = new ArrayList<>();
+        }
+        fieldDataFields.add(field);
+        return this;
+    }
+
+    public List<ScriptField> getScriptFields() {
+        return scriptFields;
+    }
+
+    public InnerHitBuilder setScriptFields(List<ScriptField> scriptFields) {
+        this.scriptFields = scriptFields;
+        return this;
+    }
+
+    public InnerHitBuilder addScriptField(String name, Script script) {
+        if (scriptFields == null) {
+            scriptFields = new ArrayList<>();
+        }
+        scriptFields.add(new ScriptField(name, script, false));
+        return this;
+    }
+
+    public FetchSourceContext getFetchSourceContext() {
+        return fetchSourceContext;
+    }
+
+    public InnerHitBuilder setFetchSourceContext(FetchSourceContext fetchSourceContext) {
+        this.fetchSourceContext = fetchSourceContext;
+        return this;
+    }
+
+    public List<SortBuilder<?>> getSorts() {
+        return sorts;
+    }
+
+    public InnerHitBuilder setSorts(List<SortBuilder<?>> sorts) {
+        this.sorts = sorts;
+        return this;
+    }
+
+    public InnerHitBuilder addSort(SortBuilder sort) {
+        if (sorts == null) {
+            sorts = new ArrayList<>();
+        }
+        sorts.add(sort);
+        return this;
+    }
+
+    public HighlightBuilder getHighlightBuilder() {
+        return highlightBuilder;
+    }
+
+    public InnerHitBuilder setHighlightBuilder(HighlightBuilder highlightBuilder) {
+        this.highlightBuilder = highlightBuilder;
+        return this;
+    }
+
+    public QueryBuilder<?> getQuery() {
+        return query;
+    }
+
+    public InnerHitBuilder setQuery(QueryBuilder<?> query) {
+        this.query = query;
+        return this;
+    }
+
+    public InnerHitBuilder setInnerHitsBuilder(InnerHitsBuilder innerHitsBuilder) {
+        this.innerHitsBuilder = innerHitsBuilder;
+        return this;
+    }
+
+    public InnerHitsContext.BaseInnerHits buildInline(SearchContext parentSearchContext, QueryShardContext context) throws IOException {
```

Comment:
The main differences are that top level inner hits can have its own query and allows for nesting. (so in case of a parent, child and grand_child scenario these can be nested inside a top level inner hits definition)

The reason I have made two different methods is that I didn't want inline inner hits logic to be complex (esp. the recursion that it has) while that isn't needed. Also top inner hits will be deprecated in 5.0 and removed later, which means this entire `buildTopLevel(...)` can simply be dropped. 

I'll add jdocs that explain this.